js/adapt-assessmentArticleModel.js

```diff
@@ -653,47 +654,44 @@ define([
 
     getSaveState: function() {
       var state = this.getState();
-      var indexByIdQuestions = [];
+      let blocks;
       var cfg = this.getConfig();
       var banksActive = cfg._banks && cfg._banks._isEnabled && cfg._banks._split.length > 1;
       var randomisationActive = cfg._randomisation && cfg._randomisation._isEnabled;
 
       if (!banksActive && !randomisationActive) {
-        // include presentation component IDs in save state so that blocks without questions aren't removed
-        this.findDescendantModels('components').forEach(function(component) {
-          var componentModel = {
-            _id: component.get('_id'),
-            _isCorrect: component.get('_isCorrect') === undefined ? null : component.get('_isCorrect')
-          };
-
-          indexByIdQuestions.push(componentModel);
-
-        });
-
-        indexByIdQuestions = _.indexBy(indexByIdQuestions, '_id');
+        // include presentation blocks in save state so that blocks without questions aren't removed
+        blocks = this.findDescendantModels('block');
       } else {
-        indexByIdQuestions = _.indexBy(state.questions, '_id');
-      }
-
-      for (var id in indexByIdQuestions) {
-        if (indexByIdQuestions.hasOwnProperty(id)) {
-          indexByIdQuestions[id] = indexByIdQuestions[id]._isCorrect;
-        }
+        blocks = state.questions.map(question => Adapt.findById(question._id).getParent());
       }
+      blocks = blocks.filter(block => {
+        const trackingId = block.get('_trackingId');
+        return Number.isInteger(trackingId) && trackingId >= 0;
+      });
+      const blockTrackingIds = blocks.map(block => block.get('_trackingId'));
+      const blockCompletion = blocks.map(block => {
+        let questions = block.findDescendantModels('question');
```

Comment:
Nice – should probably update the JSDoc comments for `AdaptModel.findDescendantModels` at some point to reflect the fact you can pass in model types other than MPABC etc.

Reply:
also could do with changing the jsdoc comments on that to remove the pluralisation of the sample values for the `descendants` parameter